### ISSUE TITLE
CBLIS - Slow down UI thread when saving CoreData context

### ISSUE THREAD

**Issue by pasin** [couchbase/couchbase-lite-ios#793]
@mastohhh reported this to me via email:

I’m @mastohhh, I'm writing directly to you because I don’t find answers on Github or Google Groups.
(I created a topic but no answer : https://groups.google.com/forum/#!topic/mobile-couchbase/SSUuoF_DJ_Q)

My question is : I have performance problems with couchbase-lite and CBLIncrementalStore, do you ? What’s the best way to create core data stack with CBLIS that doesn’t slows the UI (main thread) ?

I did a little hack to get the private context of the NSManagedObjectContext, here’s my code.

    CBLManager *manager = [[CBLManager sharedInstance] copy];
    [CBLIncrementalStore setCBLManager:manager];

    self.parentContext = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSPrivateQueueConcurrencyType];
    [self.parentContext setPersistentStoreCoordinator: coordinator];

    [self.parentContext performBlock:^{
        manager.dispatchQueue = dispatch_get_current_queue();

        NSError *error;

        self.incrementalStore = (CBLIncrementalStore*)[coordinator addPersistentStoreWithType:[CBLIncrementalStore type]
                                                                               configuration:nil
                                                                                         URL:incrementalStoreURL
                                                                                     options:nil
                                                                                       error:&error];

        self.context = [[NSManagedObjectContext alloc] initWithConcurrencyType:NSMainQueueConcurrencyType];
        self.context.parentContext = self.parentContext;

        [self.incrementalStore addObservingManagedObjectContext:self.context];
        [self.incrementalStore addObservingManagedObjectContext:self.parentContext];

        NSURL *remoteDbURL = [NSURL URLWithString:COUCHBASE_SYNC_URL];
        [self startReplication:[self.incrementalStore.database createPullReplication:remoteDbURL]];
        [self startReplication:[self.incrementalStore.database createPushReplication:remoteDbURL]];
    }]; 

Before Couchbase, this configuration was quite speed, because the context with NSMainQueueConcurrencyType didn’t write directly on the SQL file, it was his parent’s context with NSPrivateQueueConcurrencyType that did it.

Now I think there is work on the main thread with Couchbase that slows the UI after every context’s save.

I call save on every user’s actions, because I wan’t a fast sync.

What’s your strategy ? How is your core data stack ? Main queue ? Private queue ? Custom Queue ? CBLManager’s queue ?

**Comment by pasin** [comment]
I believe by default CBLIS operates on the main thread. Setting a dispatch queue to the manager given to the CBLIS should allow the CBLIS to operate on the dispatch queue thread instead. 

I think we need to do the followings:

1. Verify what thread / dispatch queue that CoreData uses to call CBLIS to execute operations. Is it the same thread as we give to the manager object or not?

2. Profile CBLIS on saving operations.

**Comment by cflorion** [comment]
I agree about the performance issues. This makes CBIS non usable in production for now (in my use case). :s
Is this more a problem of thread or a problem of too high CPU usage ?
From what I understood, CBLIS actually writes to SQLite, as regular Core Data does. What is the difference between the two that justifies the performances hit ? Revisions maybe ?

**Comment by mastohhh** [comment]
I added a pull request that performs more work on background queues : https://github.com/couchbase/couchbase-lite-ios/pull/797

**Comment by mastohhh** [comment]
Now, when i use Apple's Instruments, i only see problems in :

```objc
- (id) newValueForRelationship: (NSRelationshipDescription*)relationship
               forObjectWithID: (NSManagedObjectID*)objectID
                   withContext: (NSManagedObjectContext*)context
                         error: (NSError**)outError
```

Exactly at the call of 

```objc
// one-to-many
CBLQueryEnumerator* rows = [self queryToManyRelation: relationship
                                       forParentKeys: @[[objectID couchbaseLiteIDRepresentation]]
                                            prefetch: NO
                                            outError: outError];
```


Here's the result :

![capture d ecran 2015-06-30 a 18 37 05](https://cloud.githubusercontent.com/assets/1760064/8436243/0e175c12-1f57-11e5-9ade-793afca3ab97.png)

I don't know what to do now. I think I can't do anything on the CBLIncrementalStore stack. Maybe someone has an advice or an idea on something to look at ?

**Comment by pasin** [comment]
Here are what I have found:
1. By default, the execution to the incremental store is called on the main thread unless the NSManagedObjectContext associated with the persistent store coordinator is created with NSPrivateQueueConcurrencyType.

2. By default, CBLIS use the shared CBLManager object which operates on the main thread. For performance optimization, users could specified a new CBLManager object with a dispatch queue to the CBLIS object. The tricky part is that this needs to be done before adding persistent store to the coordinator object.

3. CBLIS's executeSaveRequest:withContext:outError: method should be wrapped inside a database transaction. This will improve performance also.

4. The couchDocumentsChanged: method is currently broken (processCouchbaseLiteChanges select will not get called) when using CBLIS with a dispatch queue. This is about mixing between dispatch queue and runloop call. Need to think a way to handle this issue.

**Comment by pasin** [comment]
@mastohhh I have made some fixes to improve the performance when executing save request. Could you please try it out and let me know? We can reopen the issue anytime.

**Comment by pasin** [comment]
Also regarding your question about core data stack, the way you have setup is fine - all of the CBLIS operation should be executed in the parentContext private dispatch queue.

**Comment by mastohhh** [comment]
Despite the optimizations @pasin did, I still encounter performance issues.

I really don't know what's the best practice to use both `CBLManager` and `NSManagedObjectContext`. Here's what I tried : 

## 1 - Parent Context - Private context's queue : 

- A main context that has a parent context
- A private context parent of the main, and with a coordinator.
- `dispatchQueue` of `CBLManager` is the same as the private context's queue. 

This is the one I actually use.

The problem with this one is `dispatch_get_current_queue();` will be deprecated. And it's not really fast.

## 2 - Parent Context - Background Queue

- A main context that has a parent context
- A private context parent of the main, and with a coordinator.
- `dispatchQueue` of `CBLManager` is `dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)`.

It gives sometimes crashes like :  `The FMDatabase <CBL_FMDatabase: 0x7ae753c0> is currently in use.`
`*** Assertion failure in -[CBL_FMDatabase warnInUse], FMDatabase.m:360`

## 3 - Main Context - Main Thread : 

- Only one context in the main thread with the coordinator.
- `dispatchQueue` of `CBLManager` is by default, main thread.

This is very slow

## 4 - Main Context - Background Queue : 

- Only one context in the main thread with the coordinator.
- `dispatchQueue` of `CBLManager` is `dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)`.

Every calls to `performBlock:` fails.
Slow and crashes like :  `The FMDatabase <CBL_FMDatabase: 0x7ae753c0> is currently in use.`

## 5 - Confinment Context - Background Queue

- `dispatchQueue` of `CBLManager` is `dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)`.
- Only one context `NSConfinementConcurrencyType` (deprecated) in the `dispatchQueue` of `CBLManager`.

Error message `Can only use -performBlock: on an NSManagedObjectContext that was created with a queue.`

## Conclusion

Performance goes down as the global number of documents increase.

Our app needs to handle thousands of documents (our firsts users maybe 100 000). There was no problem with classic CoreData SQLite. Now we have to purge old documents, and still remains problems of performance. Both in the saves, as in the fetches.

As iOS9 is available on iPad 2, we can't have an app that blocks for many seconds during usage. iPad Air 1 has UI lags from 5000 documents. 

I think pure couchbase-lite-ios without CoreData can be fast with `CBLManager`'s `dispatchQueue`, but maybe it's not design for `NSManagedObjectContext`s ?

Do you have a CoreData stack that works fine ?

Do you think `CBLIncrementalStore` can be optimized again ? I saw that `CBLQueryEnumerator* rows = [query run: outError];` is called often and behind there's `int rc = sqlite3_step(statement.statement);` and I don't think we can optimize that.

Thanks

**Comment by mastohhh** [comment]
Do I need to create a new issue ?
